Document request parameter handling

d-maurer opened the pull request:
see #636

jugmac00 commented on src/ZPublisher/request_params.rst:
```diff
@@ -0,0 +1,296 @@
+Introduction
+============
+
+Zope responds to requests, specified via an url, request headers
```
url => URL

jugmac00 commented on src/ZPublisher/request_params.rst:
```diff
@@ -0,0 +1,296 @@
+list
+  collect all values with this name into a list.
+
+  If there are more than a single parameter with the same name,
```
"If there is..." - but I am not 100% sure about this.

d-maurer replied:
I have reformulated to ensure the plural is correct.

jugmac00 commented on src/ZPublisher/request_params.rst:
```diff
@@ -0,0 +1,296 @@
+  It tells Zope to create a request variable *var*
+  with as value a 
+  record (i.e. a ``ZPublisher.HTTPRequest.record`` instance) and
+  set its attribute *attr* to the parameter value.
```
Sounds a bit rough.... maybe... "It tells Zope to create a request variable *var* of type record...

jugmac00 commented on src/ZPublisher/request_params.rst:
```diff
@@ -0,0 +1,296 @@
+1. Some variables are set up:
+
+   isFileUpload
+     does the parameter represents an uploaded file?
```
represents => represent

jugmac00 commented on src/ZPublisher/request_params.rst:
```diff
@@ -0,0 +1,296 @@
+
+As a security measure, mainly for DTML use, request variables
+are not only made available in the request attribute ``form``;
+a (somewhat) securized version of them is made available in
```
Does the word "securized" exist?

jugmac00 commented:
@d-maurer Thank you very much for this extensive documentation! I really wished this existed before before I worked on the Object Publisher part of the Zope Developer documentation, as your text really makes it very clear what happens and what are the rough edges.

I also feel your documentation offers way more value than parts of the Zope Developer documentation, and I feel it sells itself short as it is located in a package (and possibly not rendered to html) instead of in the documentation. At minimum in the long term it should get a more prominent location.

I am a bit in a rush (on my way to Pycon Web in Munich), so I could not review it in detail - but will catch up later or beginning of next week.

d-maurer commented:
@jugmac00 
> ...
> I also feel your documentation offers way more value than parts of the Zope Developer documentation, and I feel it sells itself short as it is located in a package (and possibly not rendered to html) instead of in the documentation. At minimum in the long term it should get a more prominent location.

I plan to soon propose a major overhaul of the request parameter handling - one that gets rid of most of the known bugs/caveats. The new implementation should get much clearer, too; it seems for example not necessary, that each subcase has parallel code for the tainted and untainted variant; instead construct the untainted ``form`` dict and then derive the tainted variant thereof.

dataflake commented:
Thank you Dieter, that is great documentation. You may have noticed that I did some grammar/wording changes earlier.

The only aspect that makes me unhappy is hiding it away in a place where no one will find it. With my last commit I have merged the contents into the Zope Developer Guide. Please take a look and let me know if this is OK. If yes then we can delete the document at the location in the ZPublisher folder.

d-maurer commented:
Jens Vagelpohl wrote at 2019-5-24 17:01 -0700:
>Thank you Dieter, that is great documentation. You may have noticed that I did some grammar/wording changes earlier.

I am not happy with the section title change
`Request parameter directive kinds` -->
`Request parameter directives` because the section actually describes only
the directive kinds and not the individual directives.
This is valid also for the last paragraph of the introduction.

I think english prefers `converter specific` over `converter-specific`.

The description in section "Processing model" is too detailed for
an end user documentation. You have edited out my explanation
why I have chosen this level of details (--> peculiar logig leading
to surprises): the target audience for my documentation has been
the "developers **OF** Zope", not the "developers **WITH** Zope".
In the documentation for "developers with Zope", the whole
section "Processing model" (and references thereof) can go away.

In list item 3. of section "Processing model" the `in addition` and `instead` described details of some importance: typically, only variables set up in step 1. are updated; `:tuple` updates `flags` (it ors `SEQUENCE`) and in addtion `tuple_items`; method (all variants of this kind) directives update `method` instead. The peculiar handling of `:tuple` explains the later described bug that it often is equivalent to `:list`.

In the description of the `defaults --> form` handling, a spurious `if` has slipped in.

dataflake commented:
I have incorporated your suggestions. My goal was to have the text in the Zope Developer's Guide as the only copy and remove the text file in the ZPublisher module, so if you could concentrate just on the ZDG parts that would be great.

You're correct that some of the details are unsuitable for normal application developers or Zope users and I have added a note to that effect at the start of the processing model section in the ZDG. However, we do not have any canonical place for documentation aimed at Zope developers, the ZDG is the closest we have. People won't have a hard time finding it there.

d-maurer commented:
@dataflake I am quite happy with the current state.

I have made some minor modifications:
* fixed the documentation regarding the request lookup order
* removed some items from the "converter"s list which are now classified as "aggregator"s
* used the same formatting for "aggregator"s and "converter"s.

There is still one point which does not satisfy me fully: the description is now placed into the context of object publishing, i.e. focused on the use of request variables as arguments for the call of the "published object".  Maybe, we should add a special note that this is only one use of request variables and that they are often accessed via the request object.

icemac commented on docs/zdgbook/ObjectPublishing.rst:
```diff
@@ -727,7 +759,10 @@ many converters:
 
-If the publisher cannot coerce a request variable into the type
+The full list of supported converters can be found
+in ``ZPublisher.Converters.tpye_converters``.
```
```suggestion
in ``ZPublisher.Converters.type_converters``.
```

Additionally could there be a link to the module?

icemac commented on docs/zdgbook/ObjectPublishing.rst:
```diff
@@ -951,6 +935,220 @@ simple as possible.
+
+2. The parameter value is checked to see if it is a file upload.
+   In this case, it is wrapped into a ``FileUpload``, and ``isFileUpload``
+   is updated
```
```suggestion
   is updated.
```

icemac commented on docs/zdgbook/ObjectPublishing.rst:
```diff
@@ -951,6 +935,220 @@ simple as possible.
+
+     Processing types are "ignore", "aggregate as sequence",
+     "aggregate as record", "aggregate as records", "use as default",
+     "convert" (using ``converter_type`` and ``character_encoding``)
```
```suggestion
     "convert" (using ``converter_type`` and ``character_encoding``).
```
